sky/packages/sky/lib/src/widgets/navigator2.dart

```diff
@@ -68,139 +69,68 @@ class Navigator extends StatefulComponent {
 }
 
 class NavigatorState extends State<Navigator> {
-  GlobalKey<OverlayState> _overlay = new GlobalKey<OverlayState>();
-  List<Route> _history = new List<Route>();
+  final GlobalKey<OverlayState> _overlayKey = new GlobalKey<OverlayState>();
+  final List<Route> _ephemeral = new List<Route>();
+  final List<Route> _modal = new List<Route>();
 
   void initState() {
     super.initState();
-    _addRouteToHistory(new PageRoute(
-      builder: config.routes[_kDefaultPageName],
-      name: _kDefaultPageName
-    ));
+    push(config.onGenerateRoute(new RouteArguments(name: Navigator.defaultRouteName)));
   }
 
-  RouteBuilder _generatePage(String name) {
-    assert(config.onGeneratePage != null);
-    return config.onGeneratePage(name);
+  bool get hasPreviousRoute => _modal.length > 1;
+  OverlayState get overlay => _overlayKey.currentState;
+
+  OverlayEntry get _currentOverlay {
+    for (Route route in _ephemeral.reversed) {
+      if (route.topEntry != null)
+        return route.topEntry;
+    }
+    for (Route route in _modal.reversed) {
+      if (route.topEntry != null)
+        return route.topEntry;
+    }
+    return null;
   }
 
-  bool get hasPreviousRoute => _history.length > 1;
+  Route get _currentRoute => _ephemeral.isNotEmpty ? _ephemeral.last : _modal.last;
+  Route _removeCurrentRoute() => _ephemeral.isNotEmpty ? _ephemeral.removeLast() : _modal.removeLast();
```

Comment:
I think I'd blow this up into a { } function to make it more obvious that it's not a getter, but has major side-effects. (It took me a while to find it when I was reviewing `pop()`.)